### DISCUSSION TITLE
[RFC] v5 compatible & codemod scope

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#36909] [RFCs]
# 📋 RFC

## Summary

Provider `@ant-design/compatible` package which support:
* v4-compatible-v3 (already done)
* v5-compatible-v4
* **v5-compatible-v3** (🚨 key discuss part)

## Basic example

Components:
```diff
-- import { Comment, Button } from 'antd';
++ import { Button } from 'antd';
++ import { Comment } from '@ant-design/compatible';
```

Less Varibles:
```diff
-- @import (reference) '~antd/dist/antd.less';
++ // @import (reference) '~antd/dist/antd.less';
```

## Motivation

For antd v5, we will mainly do:
* remove jsx syntactic sugar
* Align API naming standard
* migrate from less to css-in-js

Which will break current developer usage. We should provide the ability that make developer easy migration.

#### To make this RFC done:
1. Warning for v4 legacy usage as above described
2. Create the compatible package with v5 version
  * Provide one time less variable (which not allow modify)
3. Codemod for quick migration
  * Modify less file path from antd to compatible pkg
  * Modify removed component to compatible or ProComponent pkg
  * Move BackTop to FloatButton.BackTop
  * Move `antd/lib/locales` to `antd/locales`
    * Comment -> Compatible, PageHeader -> ProCoponents
  * Provide strong mode to modify props to latest support one
    * visible -> open, dropdownClassName -> popupClassName
 
#### Affect components listing:
* v5-compatible-v4
  * Dropdown: API changed
  * Menu: API changed
  * PageHeader: removed
  * Steps: remove syntactic sugar
  * AutoComplete: API changed, remove syntactic sugar
  * Checkbox: remove syntactic sugar
  * Cascader: API changed
  * DatePicker: API changed
  * Mentions: API changed, remove syntactic sugar
  * Radio: remove syntactic sugar
  * Select: API changed, remove syntactic sugar
  * TreeSelect: API changed, remove syntactic sugar
  * TimePicker: API changed
  * Comment: removed
  * Collapse: remove syntactic sugar
  * Tree: remove syntactic sugar
  * Timeline: remove syntactic sugar
  * Tag: breaking change
  * Tabs: remove syntactic sugar
  * Table: API changed, remove syntactic sugar
  * ConfigProvider: API changed

## Drawbacks

Not support real less modification since we've change to css-in-js version. Which is only to provide the file to no to break reference usage.

## Adoption strategy

First finish API updating in v4 version and PR on compatible package.

**Comment by afc163** [comment]
@ant-design/compatible 里有哪些组件？

**Reply by zombieJ** [reply]
全部组件都直接从兼容包里导出来，这样可以webpack 直接 alias。

**Reply by afc163** [reply]
1. 全部组件指什么，Button 这种没改过的也要导出么？
2. v3 里的 Form、Icon、Mention 也罗列一下。

**Reply by afc163** [reply]
@ant-design/compatible 都能 export 出哪些东西，不同版本有啥区别。

**Reply by zombieJ** [reply]
👌 我补充一下

**Comment by afc163** [comment]
How about 

```jsx
import { xxx } from '@ant-design/compatible/v3';
import { xxx } from '@ant-design/compatible/v4';
```

**Reply by zombieJ** [reply]
v5 兼容主要是在 API 上的 HOC，倒不至于拆的这么细。

**Comment by crazyair** [comment]
antd 依赖的 rc-xx 没有变更吗

**Reply by zombieJ** [reply]
变了，不过在 v4 已经做兼容式更新了。

**Reply by crazyair** [reply]
旧版如果不发版，不维护，是否可以考虑 webpack externals 方式，引入一个 antd.min.js

**Reply by zombieJ** [reply]
多版本 antd 对 bundle size 不友好，同时 antd 本身对 global 是有一些依赖，比如说 Modal 多版本弹窗那滚动条其实就同步不了。所以还是在一个包的上下文会更兼容。

**Reply by crazyair** [reply]
对于大系统，多一个 min.js 也是可以接受的，毕竟有缓存后再也不会请求，对于小页面，也没有升级负担。不需要兼容方式。

**Comment by afc163** [comment]
如何在一个包内同时兼容 v3 和 v4？

**Reply by zombieJ** [reply]
我更新到 Basic Sample 里去。

**Comment by aladdin-add** [comment]
> remove jsx syntactic sugar

这个是指 jsx/runtime: https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html?

**Reply by zombieJ** [reply]
jsx 语法糖，例如：

```tsx
<Menu>
  <Menu.Item />
</Menu>
```

to

```tsx
<Menu items={[...]} />
```

ref: https://github.com/ant-design/ant-design/issues/4853

**Comment by vagusX** [comment]
对于语法糖的话，迁移不太好处理

**Comment by vagusX** [comment]
### JSX
- [ ] Comment
- [ ] Pageheader
- [ ] Visible -> open
- [ ] Dropdown dropdownClassname -> popupClassname

**Comment by vagusX** [comment]
## Less
less import from antd 全部注释处理

**Comment by vagusX** [comment]
发现兼容包的兼容策略有两个问题
* compatible v1 时默认会引入 css `import '@ant-design/compatible/assets/index.css';`，但是 v5 的兼容包没有这个文件了，因此需要移除？
* compatible v5 没有 Mention 组件了，如果用户是从 v3 -> v5 且用了这个组件，会出现依然从 compatible 引入 Mention，从而找不到对应组件的问题，需要我们有一个说明或者提示。而默认的兼容包样式在上一条被我们移除了，可能 codemod 还需要一条针对兼容包的 Mention 组件做一条转换为引入单个 css 文件

**Reply by zombieJ** [reply]
1. 嗯，可以移除了
2. 这个组件用户用的很少，所以暂时不加了。

**Reply by vagusX** [reply]
对于第二个组件我的想法是考虑搞一个 `@ant-design/compatible-v1` 的包，仅导出 Mention 一个组件的 js 和 css，先可以不处理

**Comment by ManuRodgers** [comment]
What is wrong with Comment component? Why did you guys remove it
